20190425-keras-premade-models.md

```diff
@@ -0,0 +1,380 @@
+* distributed mini-batch training
+* accelerated training (GPU & TPU)
+* enable AutoML to select different pre-made models
+* enable users to freely compose pre-made models with other Keras layers and models.
+
+Currently most pre-made models from tensorflow have been exposed through [CannedEstimator](https://www.tensorflow.org/guide/premade_estimators). However estimator
+* relies on continuous graph rebuilding and checkpoints reloading which slows down training
+* relies on global collections and not TF 2.0 friendly.
+* makes many advanced features such as meta/transfer learning difficult
+* enforces user to create input functions when not necessary
+
+Given the status, we propose the following pre-made models as a start:
+1. Linear models including Linear Regression and Logistic Regression
+2. DNN models including DNN Regression and DNN Classification
+3. Wide-Deep models including Wide-Deep Regression and Wide-Deep Classification
+4. BoostedTree models including BoostedTree Regression and BoostedTree Classification
+
+Users can use [feature_columns](https://www.tensorflow.org/guide/feature_columns) with pre-made models for feature engineering. We use **pre-made** and **canned** interchangeably in this doc.
+
+Several side notes: This document should serve as a foundation for other pre-made models that are not included in this proposal, such as Kmeans, random forest, collaborate filtering, which are yet to come; Detailed design regarding implementation details for BoostedTree, such as distributed training, saving & restoring, acceleration, supported feature columns should live in separate docs. 
+
+## Design Proposal
+
+### Challenges
+Many canned models, including BoostedTrees, KMeans (as well as WideDeep and many others more in the future) are highly complicated and do not follow the simple foward & backprop workflow, an assumption heavily relied on by Keras. Building such models, while not compromising the basic composability of Keras (i.e., compose layers on top of layers), is the main challenge we need to resolve in this document. Another challenge is that these models can have multiple training phases (such as collaborative filtering models). We propose the following two approaches to address them:
+
+### Proposal 1: Customized training function & composability
+We propose to let each subclassed pre-made model override the training function. It is optional to provide a special subclass `CannedModel` if other methods such as `compile` and `fit` needs to be overriden as well. In traditional Keras models, such training function is dominated by autodiff - given the forward pass of the model, the gradients for each Operation is generated and backprop computation is automatically laid out for the entire model. However such assumption is only valid for neural network based supervised learning architecture. For many other scenarios, we would need to break this assumption:
+1. gradients may not be used, e.g., any un-supervised learning tasks
```

Comment:
The subclass `CannedModel` seems a like a half measure. If the training function is heavily modified it seems strange not to have an object type that describes that. Perhaps `IrregularModel` which would specify models that do not follow the typical Keras paradigm. 

The canned DNN model is a `CannedModel` imo, but not an irregular model.

Reply:
The type `CannedModel` also makes it sound as if users are not supposed to be building their own unique model types. To me canned mean pre-made.

Reply:
It's probably better not to have `CannedModel` for now, and let's just override train_function (or train_on_batch) -- WDYT?

Reply:
Agree that's better than an optional subclass. I would probably still prefer a new Model type so that it's understood how different these are than typical Keras Models, but something like `IrregularModel` may just be more bloat to the API than its worth.

Reply:
I'm keeping this as a design question to be addressed through design reviews.